chain/client/src/stateless_validation/chunk_validator/orphan_witness_pool.rs

```diff
@@ -0,0 +1,448 @@
+use std::collections::{HashMap, HashSet};
+
+use lru::LruCache;
+use near_chain_configs::default_orphan_state_witness_pool_size;
+use near_primitives::hash::CryptoHash;
+use near_primitives::stateless_validation::ChunkStateWitness;
+use near_primitives::types::{BlockHeight, ShardId};
+
+use metrics_tracker::OrphanWitnessMetricsTracker;
+
+/// `OrphanStateWitnessPool` is used to keep orphaned ChunkStateWitnesses until it's possible to process them.
+/// To process a ChunkStateWitness we need to have the previous block, but it might happen that a ChunkStateWitness
+/// shows up before the block is available. In such cases the witness is put in `OrphanStateWitnessPool` until the
+/// required block arrives and the witness can be processed.
+pub struct OrphanStateWitnessPool {
+    witness_cache: LruCache<(ShardId, BlockHeight), CacheEntry>,
+    /// List of orphaned witnesses that wait for this block to appear.
+    /// Maps block hash to entries in `witness_cache`.
+    /// Must be kept in sync with `witness_cache`.
+    waiting_for_block: HashMap<CryptoHash, HashSet<(ShardId, BlockHeight)>>,
```

Comment:
Is it really worth maintaining this lookup map in the given context? the default size limit is 25 and I don't think it makes sense to have it much higher than that (if we so behind then there is very little chance to send the endorsement in time). I think just iterating over `witness_cache` would have been more than performant enough here.

Reply:
Yeah, that's true.

After doing so much leetcode it feels very wrong to use a suboptimal algorithm, but there's no point in introducing so much complexity and `expects` when the cache holds 25 items. You're right, he rational choice is to use a naive algorithm.
Changed to a simpler implementation.

One thing I liked about the lookup map is that it provides a "safe" api - no matter what number someone passes, the cache will behave correctly. With a naive implementation using a higher value might lead to silent performance problems, so I added a warning if someone tries to create a cache with capacity of more than 128.